be/src/formats/orc/orc_chunk_reader.cpp

```diff
@@ -998,6 +1023,71 @@ static void fill_map_column_with_null(orc::ColumnVectorBatch* cvb, ColumnPtr& co
     }
 }
 
+static void fill_struct_column(orc::ColumnVectorBatch* cvb, ColumnPtr& col, int from, int size,
+                               const TypeDescriptor& type_desc, const OrcMappingPtr& mapping, void* ctx) {
+    auto* orc_struct = down_cast<orc::StructVectorBatch*>(cvb);
+    auto* col_struct = down_cast<StructColumn*>(col.get());
+
+    Columns& field_columns = col_struct->fields_column();
+
+    size_t child_index = 0;
+    for (size_t i = 0; i < type_desc.children.size(); i++) {
+        if (!type_desc.selected_fields[i]) {
+            continue;
+        }
+
+        const TypeDescriptor& field_type = type_desc.children[i];
+        size_t column_id = mapping->get_column_id_or_child_mapping(i).orc_column_id;
+
+        orc::ColumnVectorBatch* field_cvb = orc_struct->fieldsColumnIdMap[column_id];
+        const FillColumnFunction& fn_fill_elements = find_fill_func(field_type.type, true);
+        fn_fill_elements(field_cvb, field_columns[child_index], from, size, field_type,
+                         mapping->get_column_id_or_child_mapping(i).orc_mapping, ctx);
+        child_index++;
+    }
+
+    DCHECK_EQ(field_columns.size(), child_index);
+}
+
+static void fill_struct_column_with_null(orc::ColumnVectorBatch* cvb, ColumnPtr& col, int from, int size,
+                                         const TypeDescriptor& type_desc, const OrcMappingPtr& mapping, void* ctx) {
+    auto* orc_struct = down_cast<orc::StructVectorBatch*>(cvb);
+    auto* col_nullable = down_cast<NullableColumn*>(col.get());
+    auto* col_struct = down_cast<StructColumn*>(col_nullable->data_column().get());
+
+    if (!orc_struct->hasNulls) {
+        fill_struct_column(cvb, col_nullable->data_column(), from, size, type_desc, mapping, ctx);
+        col_nullable->null_column()->resize(col_struct->size());
+    } else {
+        const int end = from + size;
+        int i = from;
+
+        while (i < end) {
+            int j = i;
+            // Loop until NULL or end of batch.
```

Comment:
Could this code be simplified? there are duplicated code down there.

Reply:
We decide to optimize the fill functions when we implement ORC lazy load.